### PR TITLE
Fixed memory issues in linalg_lstsq

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85239
* __->__ #85357

### CODE REVIEW

**Comment by ezyang** [comment]
For testing, we could just enable python dispatcher in the same way you did in the previous PR, since that was enough to trigger the problem. It may also uncover other problems this way.

**Review by ezyang — APPROVED** [review]

**Comment by IvanYashchuk** [comment]
Ref. https://github.com/pytorch/pytorch/pull/85239#issuecomment-1252713393:
> `enable_python_dispatcher` enables a boxed calling convention. Is there any other way to enable this?
> A variant of what's in https://github.com/pytorch/pytorch/pull/85239#pullrequestreview-1113249053 is good enough for a test (the result should match with and without boxed calling convention enabled).

**Comment by lezcano** [comment]
I don't understand what's going on here, perhaps @IvanYashchuk can explain. Why is `solution` not of the right size and why do we need to do a narrow? 

I think that:
- `solution` should be of the right size (no narrowing should be needed)
- If we need more space to do the computations, we should allocate a different tensor, use it for the computations, and then copy the result from the larger tensor onto `solution`
- If we need less space, we can write into `solution` directly through a view of it.

**Comment by IvanYashchuk** [comment]
> If we need more space to do the computations, we should allocate a different tensor, use it for the computations, and then copy the result from the larger tensor onto solution

It's an example of bad micro-optimization to avoid copies. Lapack interface requires a larger matrix to be passed to store both raw residuals and solution in the same matrix.

File: test/test_python_dispatch.py [hunk 1]
```diff
@@ -1839,5 +1839,13 @@ def test_basic(self):
         r = torch._C._EnablePythonDispatcher()
         torch.add(x, x)
 
+    def test_lstsq(self):
+        a = torch.randn(4, 3)
+        b = torch.rand(4, 3)
+        expected_shape = torch.linalg.lstsq(a, b).solution.shape
+        r = torch._C._EnablePythonDispatcher()
```
Comment (IvanYashchuk) [inline test/test_python_dispatch.py:1846]: Should it be wrapped in `try: finally: del r`? I'm not sure if there's a possibility to leak it into other tests.

**Review by IvanYashchuk — APPROVED** [review]
Thank you for the fix and for opening a separate PR!

**Comment by ezyang** [comment]
> It's an example of bad micro-optimization to avoid copies. Lapack interface requires a larger matrix to be passed to store both raw residuals and solution in the same matrix.

Actually, then maybe the kernel as a whole is wrong. If we pass in a preexisting out tensor, you can't assume that you get to have extra slop after the end of your tensor to store the extra info. So you have to do a copy in this case. Is that what the copy logic does?

File: test/test_python_dispatch.py (hunk 1 above)
Comment (Chillee) [reply test/test_python_dispatch.py:1846]: I think it's an RAII object, see `enable_python_dispatch`: https://github.com/pytorch/pytorch/blob/master/torch/_dispatch/python.py#L16

**Comment by IvanYashchuk** [comment]
> > It's an example of bad micro-optimization to avoid copies. Lapack interface requires a larger matrix to be passed to store both raw residuals and solution in the same matrix.
> 
> Actually, then maybe the kernel as a whole is wrong. If we pass in a preexisting out tensor, you can't assume that you get to have extra slop after the end of your tensor to store the extra info. So you have to do a copy in this case. Is that what the copy logic does?

No, if we pass in a preexisting out tensor with an unfriendly shape for Lapack then a copy is made.

**Comment by lezcano** [comment]
Note that in QR we write this pattern properly:
https://github.com/pytorch/pytorch/blob/2a88f1b2d86a1fdc7380db768e67b18c24d199c4/aten/src/ATen/native/BatchLinearAlgebra.cpp#L2307-L2319
I should port lstsq to be a structured kernel one of these days, as it keeps showing up in issues.

**Pull request closed without merging**